Initial benchmarking support for the farmer

i1i1 opened the pull request:
This pr introduces the base for future work on benchmarking. The main purpose of this pr is to:
- introduce benchmarking mock rpc
- introduce command

Creation of some benchmark introspection and statistics is out of scope here.

Things to think about:
- Should we remove cli arguments for:
 - Writing actual plot
 - `write_pieces_size` (calculate it using other arguments)

i1i1 committed:
Generalize multifarming over client

i1i1 committed:
Make farming optional

i1i1 committed:
Add mock rpc for benchmarking

i1i1 committed:
Introduce multifarming options for creation

i1i1 committed:
Add bench command

i1i1 committed:
Add option for mocking plot

i1i1 committed:
End benchmarking at some point

i1i1 committed:
Wipe plot after benchmarking

nazar-pc commented on crates/subspace-farmer/src/multi_farming.rs:
```diff
@@ -46,7 +45,7 @@ impl MultiFarming {
     pub async fn new(
         base_directory: PathBuf,
-        client: NodeRpcClient,
+        client: impl RpcClient,
```
Let's not use `impl`, use generic parameter instead

nazar-pc commented on crates/subspace-farmer/src/multi_farming.rs:
```diff
@@ -79,6 +80,7 @@ impl MultiFarming {
         best_block_number_check_interval: Duration,
         plot_sizes: Vec<u64>,
         new_plot: impl Fn(usize, PublicKey, u64) -> Result<Plot, PlotError> + Clone + Send + 'static,
+        start_farmings: bool,
```
Now abstraction seems a bit broken if we have `MultiFarming` with farming disabled. It doesn't seems to make a lot of sense when combined together, does it?

i1i1 replied:
I would leave it under the same abstraction though, as it allows code reuse. Should I rename `MultiFarming` to smth else?

nazar-pc replied:
Maybe, but I can't come up with much better name right now.

i1i1 replied:
I think this is just an exception for the benchmarking case. That's why this flag was hidden in the first place, so that you either call `new` or `benchmarking`.

nazar-pc replied:
We can keep it for now then

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -80,11 +80,13 @@ pub(crate) async fn farm(
-        object_mappings.clone(),
-        reward_address,
-        best_block_number_check_interval,
+        multi_farming::Options {
```
I'd import `Options` instead of call it more verbose if you're concerned with generic name

i1i1 replied:
I don't see how it would solve the problem if I would call it `MultiFarmingOptions`. It is still would be generic. Main difference would be in importing vs not importing.

I think that calling things with generic names inside the module and using them with this name is fine, as we know the context. Overall it is fine as it can be referred using the module name like here.

nazar-pc replied:
Then rename during import. I'm just trying to have more consistency here. We typically import data structures unless there are some conflicts or other reasons to not import. I just don't see why all of a sudden explicit path is needed here.

nazar-pc commented on crates/subspace-farmer/src/bench_rpc_client.rs:
```diff
@@ -0,0 +1,201 @@
+use std::sync::Arc;
```
Are you sure this should even be a part of the library? I think you could have added this to the CLI crate instead and not expose it from the library.

i1i1 replied:
I thought it would be usefull to reuse it from the gui, no?

i1i1 replied:
We can make it separate crate though

nazar-pc replied:
Let's just move it from library into binary for now. We have created `RpcClient` specifically such that we can implement it externally if need be.

nazar-pc commented on crates/subspace-farmer/src/bench_rpc_client.rs:
```diff
@@ -0,0 +1,201 @@
+
+                segment_index += 1;
+            }
+        });
```
Should these be a part of the client itself though? It seems to be that these should be fed into the client from the outside, so you can control it if need be.

i1i1 replied:
I can't imagine a good API apart from asking 2 closures from user which would generate segments and slot information. Do you think its fine?

nazar-pc replied:
I think it'll be better, maybe even two channels, depends on how you implement it

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -158,6 +160,74 @@ pub(crate) async fn farm(
+            };
+            log::warn!("Failed to increase file limit: {err}")
+        }
+    }
```
Worth having a utility function for this

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/main.rs:
```diff
@@ -61,6 +61,26 @@ enum Command {
     },
     /// Start a farmer using previously created plot
     Farm(FarmingArgs),
+    /// Benchmark disk in order to see a throughput of the disk for farming
```
For farming or plotting?

i1i1 replied:
Plotting, right :)

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/main.rs:
```diff
@@ -80,6 +80,9 @@ enum Command {
         max_plot_size: Option<u64>,
+        /// Actually write pieces to disk while benchmarking (might be more accurate)
+        #[clap(long)]
+        plot: bool,
```
This is not a great name. I can imagine something like `write_to_disk` that is an enum with variants that allow writing nothing, everything, just plot, just auxiliary databases, etc. Can start with this:
```
enum WriteToDisk {
    Everything,
    Nothing,
}

nazar-pc commented on crates/subspace-farmer/src/multi_farming.rs:
```diff
@@ -76,23 +77,41 @@ impl MultiFarming {
         options: Options<impl RpcClient>,
         total_plot_size: u64,
         max_plot_size: u64,
+        mock_plot: bool,
```
I think plot creation factory method would work better here, then you don't have to have `BenchPlotMock` and similar things polluting library code and can be moved to CLI.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -227,6 +228,12 @@ pub(crate) async fn bench(
     .await?;
 
+    while client.writen_piece_count() < write_pieces_size / PIECE_SIZE as u64 {
+        tokio::time::sleep(Duration::from_secs(1)).await;
```
I'm not a fan of `sleep`s, usually indication of other issues. I think moving code from `BenchRpcClient` out would allow you to wait for precise number of pieces you want to write instead of doing hacks like this.

i1i1 replied:
It is removed in the following prs. This is just a stub for now.

nazar-pc replied:
All sleeps removed in following PRs? I've seen some in the RPC client too.

i1i1 replied:
No, only this one. There is one sleep in the bench rpc client, but it is for salt changing, which I decided to make time relative. Should I make it segment or piece relative?

I think we can decide on smth relative to the api of bench rpc client in this thread also:
https://github.com/subspace/subspace/pull/442#discussion_r873085837

nazar-pc replied:
Yes, we should avoid timeouts in these things such that they finish as soon as they can and don't depend on machine speed (making tests/benches hard to reproduce).

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -234,7 +237,11 @@ pub(crate) async fn bench(
-    multi_farming.wait().await
+    multi_farming.wait().await?;
+
+    wipe(base_directory)?;
```
`TempDir` has `Drop` impl that will delete created directory once dropped, so no need to wipe it explicitly here.

nazar-pc replied:
Also I don't quite get why you need to crate subdirectory in the first place. You could have used `wipe` on original directly as it only removes known files and directories, I'm a bit confused here.

i1i1 replied:
> `TempDir` has `Drop` impl that will delete created directory once dropped, so no need to wipe it explicitly here.

True, but I just thought that it would be better to clean directory explicitly and then just remove directory itself via drop. Will remove `wipe` call then



> Also I don't quite get why you need to crate subdirectory in the first place. You could have used `wipe` on original directly as it only removes known files and directories, I'm a bit confused here.

We need to create another directory as our original directory might have some plot stored in there, so we don't want to reuse and pollute it.

i1i1 committed:
Apply minor suggestions

i1i1 committed:
Change cli flag plot mocking

i1i1 committed:
Ask for plot factory in `MultiFarming`

i1i1 committed:
Move the bench rpc client to the binary

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/main.rs:
```diff
@@ -50,6 +50,13 @@ struct FarmingArgs {
 }
 
+#[derive(Debug, Clone, Copy, ArgEnum)]
+#[clap(default_value = "nothing")]
```
Just curious, doesn't Clap support `Default`?

i1i1 replied:
Yeah, it does. Will change it :)

i1i1 replied:
Right now it doesn't work as intended

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -88,14 +128,22 @@ pub(crate) async fn farm(
         },
         plot_size,
         max_plot_size,
+        move |plot_index, address, max_piece_count| {
```
Is it an address or a public key of internal identity?

i1i1 replied:
`subspace_core_primitives::PublicKey`

nazar-pc replied:
I mean does it serve the role of public key or reward address? If it is just a public key, then should be `public_key` variable to keep things consistent.

i1i1 replied:
Yeah, it is just a public key. I will change it, thanks :)

i1i1 committed:
Apply more minor suggestions

i1i1 committed:
Change API of the benchmark rpc client.

Right now it will use channels, as it would allow to spawn your own
tasks outside which would bring more flexibility.

i1i1 committed:
Remove slot info sending, as it is unused for now

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/main.rs:
```diff
@@ -52,12 +52,17 @@ struct FarmingArgs {
 
+impl Default for WriteToDisk {
+    fn default() -> Self {
+        Self::Nothing
```
I'd expect benchmarking to write everything by default :slightly_smiling_face:

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/bench_rpc_client.rs:
```diff
@@ -32,114 +26,53 @@ pub struct Inner {
-    pub fn writen_piece_count(&self) -> u64 {
-        self.inner.writen_piece_count.load(Ordering::Relaxed)
+        };
+        (me, outer_archived_segments_sender, outer_slot_info_sender)
```
It could have received receiver side of the channel instead, then there would be no need to return multiple things here.

nazar-pc reviewed (dismissed):
Fine overall

i1i1 committed:
Apply more minor suggestions

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -156,6 +217,132 @@ pub(crate) async fn farm(
+    max_plot_size: 100 * 1024 * 1024 * 1024 / PIECE_SIZE as u64, // 100G
+};
+
+pub(crate) async fn bench(
```
Looks like I missed, but what is `fn bench` doing in `farm.rs`? It should have been `bench.rs`, the file was created specifically for `farm` command because it was non-trivial.
